### DISCUSSION TITLE
API InlineCompletions Finalised

### DISCUSSION THREAD

**Discussion by isidorn** [microsoft/vscode-discussions#26] [Announcements]
We have finalized the Inline Completions API. This allows extensions to provide inline completions that are decoupled from the suggestion widget. An inline completion is rendered as if it was already accepted, but with a grey color. Users can cycle through suggestions and accept them with the `Tab` key. 

An example extension that uses Inline Completions is GitHub Copilot. 

More information can be found in the `vscode.d.ts` file with the entrypoint into the API being [`languages.registerInlineCompletionItemProvider`](https://github.com/microsoft/vscode/blob/e3a8e502ad7263836d0bc34cbcefbfc7bd65104f/src/vscode-dts/vscode.d.ts#L12357).

Try out this API and we are very excited to see what you will build with it. Feedback is very much appreciated!

fyi @hediet

**Comment by PEZ** [comment]
Sounds very cool! Do we have some guidance somewhere around when to use which type of completion?

**Reply by hediet** [reply]
We generally recommend to use the normal completion item provider (it is much better to explore completions). There also is a user-setting to show completion items as inline completions.

**Comment by PEZ** [comment]
My extension has default bindings for both `tab` and `esc` with `when` clauses to not interfere with suggest widgets and the like. It doesn't work with these inline suggestions because I can't find a [when context](https://code.visualstudio.com/api/references/when-clause-contexts) to guard with. Is there such a context? If not, is there some other way I can disable my bindings when inline suggestions are showing?

Tracking this on Calva here: https://github.com/BetterThanTomorrow/calva/issues/1781

**Reply by PEZ** [reply]
I think I have figured it out. The default inline suggest keybindings uses this when context for accepting a suggestion with `tab`:

```
inlineSuggestionHasIndentationLessThanTabSize && inlineSuggestionVisible && !editorTabMovesFocus
```

For dismissing, simply `inlineSuggestionVisible` is used.

**Comment by Qinyun718** [comment]
Hi, Could I know when the inline completion api would be available in the official vscode (not the insider version)? Thank you!

**Reply by isidorn** [reply]
It is available in VS Code stable since 3 weeks.

**Comment by juli1** [comment]
How may I start using this API? Is there any specific changes to do in my `package.json` or any dependency to start using it?

Is there any example of how to use it?

**Reply by isidorn** [reply]
You can just start using it. 
Here's a a sample https://github.com/microsoft/vscode-extension-samples/tree/main/inline-completions

**Reply by juli1** [reply]
Thanks for the link! I tried it on the current release of vscode 1.68.1 and it does not seem to work. Looking at the repository, it seems it works only with the insider release. Any idea when it will work with the current version of vscode?

Thanks!

**Reply by isidorn** [reply]
@juli1 it should work with 1.68.1. If it does not that seems like a bug and can you please file a new issue in the VS Code github repo with reproducable steps. Thank you

**Reply by juli1** [reply]
> @juli1 it should work with 1.68.1. If it does not that seems like a bug and can you please file a new issue in the VS Code github repo with reproducable steps. Thank you

I just imported the examples and the plugin shows multiple errors when starting. This seems to be a plugin that still relies on API proposals since the `package.json` file defines `enabledApiProposals`. Even after removing the `enabledApiProposals`, the message is still there. This is what is happening when launching the example.

![Screen Shot 2022-07-06 at 3 40 09 PM](https://user-images.githubusercontent.com/993972/177648266-ac43715f-bfe0-4cae-b04e-b58a0366b86d.png)


That is kind of sad to see how this feature is being released when it was being used by GitHub for a brand new feature and not being available to all the rest of the developer ecosystem. And still lack documentation and support as of today.

I will file a bug.  What repository should I file an issue with? Any link to the repository?

Other question: it seems that any extension that wants to use the inline completion has to use [vscode-dts](https://github.com/microsoft/vscode-dts), adding more steps in the `package.json`. Any plan for when `@types/vscode` will contain the new signatures? Using `vscode-dts` is totally understandable when a API was proposed, it would be good to be able to not use it since this is finalized and released in production releases.

Thanks,

**Reply by juli1** [reply]
Bumping this thread - any update on this?

**Reply by isidorn** [reply]
Sorry for the slow response I was on vacation. Please file an issue here https://github.com/microsoft/vscode/issues
Thank you

**Reply by talbright** [reply]
Bump. This looks like its still in a proposed state? https://github.com/microsoft/vscode/blob/main/src/vscode-dts/vscode.proposed.inlineCompletionsAdditions.d.ts 

What are the blockers that remain?

**Comment by nandooliveira** [comment]
Is it possible to listen for events when an inline completion was accepted, even from a different extension?

**Reply by nandooliveira** [reply]
@isidorn

**Reply by isidorn** [reply]
I think not.

**Comment by heaversm** [comment]
I still get this error in my extension when I package it and run it as a VSIX. 

If I include 

```
  "enabledApiProposals": [
    "inlineCompletionsAdditions"
  ],
  ```
  
  In package json it works by adding my extension to arg.v or launching code with a flag, but otherwise says:
  
  `  ERR Extension 'Mozilla-Ocho.bosque CANNOT USE these API proposals 'inlineCompletionsAdditions'. You MUST start in extension development mode or use the --enable-proposed-api command line flag`.
  
  If I take it out of package.json, I get this error:
  
  ```
  mainThreadExtensionService.ts:107 Activating extension 'Mozilla-Ocho.bosque' failed: Extension 'Mozilla-Ocho.bosque' CANNOT use API proposal: inlineCompletionsAdditions.
Its package.json#enabledApiProposals-property declares: [] but NOT inlineCompletionsAdditions.
 The missing proposal MUST be added and you must start in extension development mode or use the following command line switch: --enable-proposed-api Mozilla-Ocho.bosque.
 ```

From installed vsix location extensions.json

```
{"identifier":{"id":"[PUBLISHER.PRODUCT]"},"version":"0.0.5","location":{"$mid":1,"fsPath":"/Users/mheavers/.vscode/extensions/[PUBLISHER-PRODUCT]-0.0.5","external":"file:///Users/mheavers/.vscode/extensions/[PUBLISHER-PRODUCT]-0.0.5","path":"/Users/mheavers/.vscode/extensions/ho.bosque-0.0.5","scheme":"file"},"relativeLocation":"[PUBLISHER-PRODUCT]-0.0.5","metadata":{"installedTimestamp":1734541536632,"pinned":true,"source":"vsix"}}
```

VSCODE VERSION:

```

Version: 1.96.0 (Universal)
Commit: 138f619c86f1199955d53b4166bef66ef252935c
Date: 2024-12-11T02:29:09.626Z (1 wk ago)
Electron: 32.2.6
ElectronBuildId: 10629634
Chromium: 128.0.6613.186
Node.js: 20.18.1
V8: 12.8.374.38-electron.0
OS: Darwin arm64 24.1.0
```

**Reply by suspicious-pineapple** [reply]
hello mysterious stranger from 2024, did you ever end up fixing the problem?

**Reply by PEZ** [reply]
DenverCoder9 vibes! https://xkcd.com/979/
